src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.cc

```diff
@@ -0,0 +1,180 @@
+  uint8_t checksum_bytes = (frame->checksum_present) ? 4 : 0;
+
+  // Get the section(s) data from the buffer.
+  while (decoder->BufSize() > checksum_bytes) {
+    mongodb::Section section;
+    PX_ASSIGN_OR(section.kind, decoder->ExtractLEInt<uint8_t>(), return ParseState::kInvalid);
+    // Length of the section still remaining in the buffer.
+    int32_t remaining_section_length = 0;
+
+    if (section.kind == 0) {
+      // Check the length but don't extract it since the later logic requires the buffer to retain
+      // it.
+      section.length = utils::LEndianBytesToInt<int32_t, 4>(decoder->Buf());
+      if (section.length < kSectionLengthSize) {
+        return ParseState::kInvalid;
+      }
+      remaining_section_length = section.length;
+
+    } else if (section.kind == 1) {
+      PX_ASSIGN_OR(section.length, decoder->ExtractLEInt<uint32_t>(), return ParseState::kInvalid);
+      if (section.length < kSectionLengthSize) {
+        return ParseState::kInvalid;
+      }
+
+      // Get the sequence identifier (command argument).
+      PX_ASSIGN_OR(std::string_view seq_identifier, decoder->ExtractStringUntil('\0'),
+                   return ParseState::kInvalid);
+      // Make sure the sequence identifier is a valid OP_MSG kind 1 command argument.
+      if (seq_identifier != "documents" && seq_identifier != "updates" &&
+          seq_identifier != "deletes") {
+        return ParseState::kInvalid;
+      }
+      remaining_section_length = section.length - kSectionLengthSize - seq_identifier.length() - 1;
+
+    } else {
+      return ParseState::kInvalid;
+    }
+
+    // Extract the document(s) from the section and convert it from type BSON to a JSON string.
+    auto parse_until = decoder->BufSize() - remaining_section_length;
+    while (decoder->BufSize() > parse_until) {
+      // We can't extract the length bytes since bson_new_from_data() expects those bytes in
+      // the data as well as the expected length in another parameter.
+      auto document_length = utils::LEndianBytesToInt<int32_t, 4>(decoder->Buf());
+      if (document_length > kMaxBSONObjSize) {
+        return ParseState::kInvalid;
+      }
+
+      PX_ASSIGN_OR(auto section_body, decoder->ExtractString<uint8_t>(document_length),
+                   return ParseState::kInvalid);
+
+      // Check if section_body contains an empty document.
+      if (section_body.length() == kSectionLengthSize) {
+        section.documents.push_back("");
+        continue;
+      }
+
+      bson_t* bson_doc = bson_new_from_data(section_body.data(), document_length);
+      DEFER(bson_destroy(bson_doc));
+      if (bson_doc == NULL) {
+        return ParseState::kInvalid;
+      }
+      char* json = bson_as_canonical_extended_json(bson_doc, NULL);
+      DEFER(bson_free(json));
+      if (json == NULL) {
+        return ParseState::kInvalid;
+      }
+
+      // Find the type of command argument from the kind 0 section.
+      if (section.kind == 0) {
```

Comment:
Not sure if there is any semantic description of the section kind, but if so (i.e. in particular, if so in upstream), then we could have something like this:
```
if (section.kind == kBSONSection)
```